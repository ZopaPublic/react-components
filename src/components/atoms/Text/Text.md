### Summary

Whenever you want to render text on any UI at Zopa, use the `<Text />` component.

Apply margin/padding by [adding spacing](/#/Content?id=spacing).

### Examples

- Variations

```tsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <Text as="p" className="mb-6">
    Long paragraph, rendered as `p`: we give our customers a fair deal as standard and our products are built so that we
    win when you win. Managing your money is no sweat. With handy tools to get stuff done and helpful people always at
    the end of a phone. We listen to what you want and change with your needs.
  </Text>
  <Text>Text as `span`</Text>
</Fragment>;
```

- Sizes

```tsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <Text size="lead" className="mb-6" as="p">
    Lead ( 18px )
  </Text>
  <Text className="mb-6" as="p">
    Medium ( default: 16px )
  </Text>
  <Text size="small">Small ( 14px )</Text>
</Fragment>;
```

- Weights

```tsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <Text className="mb-6" as="p">
    Regular
  </Text>
  <Text weight="bold" as="p">
    Bold
  </Text>
  <Text weight="semiBold" as="p">
    SemiBold
  </Text>
</Fragment>;
```

- Colors

```tsx { "props": { "style": { "backgroundColor": "rgb(244, 248, 246)", "border": "none" } } }
import { Fragment } from 'react';
import { Text, colors } from '@zopauk/react-components';

<Fragment>
  <Text color={colors.white} className="mb-6" as="p">
    White
  </Text>
  <Text color={colors.grey} className="mb-6" as="p">
    ❗️`grey` has poor contrast (not AAA compliant). Use it for a text that doesn't convey a critical message.
  </Text>
  <Text color={colors.greyDarkest} className="mb-6" as="p">
    Grey Darkest
  </Text>
  <Text color={colors.success} className="mb-6" as="p">
    Success
  </Text>
  <Text color={colors.alert} className="mb-6" as="p">
    Alert
  </Text>
  <Text color={colors.alertDark}>Alert Dark</Text>
</Fragment>;
```

- Alignment

```tsx
import { Fragment } from 'react';
import { Text } from '@zopauk/react-components';

<Fragment>
  <div style={{ textAlign: 'right' }}>
    <Text className="mb-6">Inherits from his parent by default</Text>
  </div>
  <Text align="left" className="mb-6" as="p">
    Left
  </Text>
  <Text align="center" className="mb-6" as="p">
    Center
  </Text>
  <Text align="right" className="mb-6" as="p">
    Right aligned
  </Text>
</Fragment>;
```

- Capitalised

```tsx
import { Text } from '@zopauk/react-components';

<Text capitalize>Vivacious</Text>;
```
