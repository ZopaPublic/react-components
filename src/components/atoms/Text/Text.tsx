import React, { HTMLAttributes } from 'react';
import styled from 'styled-components';
import { useThemeContext, AppTheme } from '../../styles/Theme';

export interface TextProps extends HTMLAttributes<HTMLSpanElement> {
  children?: React.ReactNode;
  /**
   * The weight of the rendered text.
   * @default 'regular'
   */
  weight?: 'regular' | 'bold' | 'semiBold';
  /**
   * The size you want to render your text at, currently only `13px` | `15px` and `18px` supported.
   * @default 'body'
   */
  size?: keyof AppTheme['typography']['text']['sizes'];
  /**
   * The HTML5 tag you want to render your text on, currently only `<span>` and `<p>` are supported.
   * @default 'span'
   */
  as?: 'span' | 'p' | 'figcaption';
  /**
   * Whether to render the text in all caps or not.
   * @default false
   */
  capitalize?: boolean;
  /**
   * Where the rendered text should be aligned to.
   * @default 'inherit'
   */
  align?: 'inherit' | 'left' | 'right' | 'center';
  /**
   * Accepts a subset of the Zopa brand colors.
   * @default `colors.greyDarkest`
   */
  color?: string;
}

const Text = styled.span<TextProps & { theme: AppTheme }>`
  margin: 0;
  letter-spacing: ${({ theme }) => theme.typography.letterSpacingMap.text};

  color: ${({ color, theme }) => color || theme.typography.text.color};
  font-size: ${({ theme, size = 'body', capitalize }) => theme.typography.text.sizes[capitalize ? 'small' : size]};
  line-height: ${({ theme, size = 'body' }) => theme.typography.lineHeight[size]};
  font-weight: ${({ theme, weight = 'regular', capitalize }) => {
    return theme.typography.weights[capitalize ? 'bold' : weight];
  }};

  font-family: ${({ theme }) => theme.typography.primary};
  text-align: ${({ align = 'inherit' }) => align};
  text-transform: ${({ capitalize }) => capitalize && 'uppercase'};

  ${({ className = '', as }) =>
    (!as || as === 'span') &&
    className.split(' ').some((clss) => /[mp](:[mp])?[tblrxy]?-\d+/.test(clss)) &&
    `
    display: block
  `}
`;

const TextWrap = React.forwardRef<HTMLSpanElement, TextProps>((props, ref) => {
  const theme = useThemeContext();
  return <Text {...props} ref={ref} theme={theme} />;
});

export default TextWrap;
