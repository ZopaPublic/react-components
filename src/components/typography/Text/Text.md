Text component

### Text sizes

```jsx
import { Text } from '@zopauk/react-components';

<>
  <div>
    <Text size="xl">Extra large</Text>
  </div>
  <div>
    <Text size="l">Large</Text>
  </div>
  <div>
    <Text size="m">Medium (default)</Text>
  </div>
  <div>
    <Text size="s">Small</Text>
  </div>
  <div>
    <Text size="xs">Extra small</Text>
  </div>
</>;
```

### Text font weights

```jsx
import { Text } from '@zopauk/react-components';

<>
  <div>
    <Text fw="normal">Normal (default)</Text>
  </div>
  <div>
    <Text fw="bold">Bold</Text>
  </div>
</>;
```

### Text custom colors

```jsx
import { Text } from '@zopauk/react-components';

<>
  <div>
    <Text>Default</Text>
  </div>
  <div>
    <Text color="#00B9A7">Teal</Text>
  </div>
  <div>
    <Text color="#5E35B1">Indigo</Text>
  </div>
</>;
```

### Text as different html elements

```jsx
import { Text } from '@zopauk/react-components';

<>
  <div>
    <Text as="span">Text as span tag (default)</Text>
  </div>
  <div>
    <Text as="p">Text as p tag</Text>
  </div>
</>;
```
