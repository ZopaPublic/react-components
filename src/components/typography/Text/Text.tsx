import React from 'react';
import styled from 'styled-components';
import * as colors from '../../../constants/colors';
import * as fonts from '../../../constants/fonts';

const sizes = {
  l: '18px',
  m: '16px',
  s: '14px',
  xl: '20px',
  xs: '12px',
};

export type TTextSize = keyof typeof sizes;

const fontWeights = {
  bold: 600,
  normal: 400,
};

export type TTextFontWeight = keyof typeof fontWeights;

export interface IStyledTextProps extends React.HTMLAttributes<HTMLSpanElement | HTMLParagraphElement> {
  color: string;
  fw: TTextFontWeight;
  size: TTextSize;
}

const StyledSpan = styled.span<IStyledTextProps>`
  margin: 0;
  letter-spacing: 0;
  line-height: 24px;
  font-family: ${fonts.openSans};
  font-weight: ${({ fw }) => fontWeights[fw]};
  font-size: ${({ size }) => sizes[size]};
  color: ${({ color }) => color};
`;

export type TTextTag = 'span' | 'p';

export interface ITextProps extends React.HTMLAttributes<HTMLSpanElement | HTMLParagraphElement> {
  /**
   * font weight
   * @default 400
   */
  fw?: TTextFontWeight;
  /**
   * size
   * @default 14px
   */
  size?: TTextSize;
  /**
   * color
   * @default colors.neutral.neutral900
   */
  color?: string;
  /**
   * rendered tag
   * @default span
   */
  as?: TTextTag;
}

const Text = ({ fw = 'normal', size = 'm', color = colors.neutral.neutral900, as = 'span', ...rest }: ITextProps) => (
  <StyledSpan fw={fw} size={size} color={color} as={as} {...rest} />
);

export default Text;
