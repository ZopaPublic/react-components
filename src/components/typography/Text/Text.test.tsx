import { axe } from 'jest-axe';
import React from 'react';
import { render } from 'react-testing-library';
import Text, { ITextProps } from './Text';

describe('Text', () => {
  it('renders the component with no a11y violations', async () => {
    const { container } = render(<Text>Text</Text>);
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });

  it('renders the component default props', () => {
    const { container } = render(<Text>Text</Text>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component custom props', () => {
    const props: ITextProps = {
      color: '#111',
      fw: 'bold',
      as: 'p',
    };
    const { container } = render(<Text {...props}>Text</Text>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
