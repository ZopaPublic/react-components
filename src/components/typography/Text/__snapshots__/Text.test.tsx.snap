// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders the component custom props 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 24px;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 600;
  font-size: 16px;
  color: #111;
}

<p
  class="c0"
  color="#111"
>
  Text
</p>
`;

exports[`Text renders the component default props 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 24px;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 400;
  font-size: 16px;
  color: #1C2139;
}

<span
  class="c0"
  color="#1C2139"
>
  Text
</span>
`;
